File: app/src/main/java/com/originb/inkwisenote2/common/ListUtils.java
package com.originb.inkwisenote2.common;

import java.util.*;
import java.util.function.BinaryOperator;
import java.util.function.Function;
import java.util.stream.Collectors;

public class ListUtils {
    public static <T> List<T> listOf(T... items) {
        if (items == null) return new ArrayList<>();
        return new ArrayList<>(Arrays.asList(items));
    }

    public static <T> List<T> merge(List<T> a, List<T> b) {
        if (a == null && b == null) return new ArrayList<>();
        else if (a != null && b == null) return new ArrayList<>(a);
        else if (a == null && b != null) return new ArrayList<>(b);

        List<T> result = new ArrayList<>(a);
        result.addAll(b);
        return result;
    }

    public static <T, K> Map<K, List<T>> groupBy(List<T> list, Function<? super T, ? extends K> keyMapper) {
        return list.stream().collect(Collectors.<T, K, List<T>>toMap(keyMapper, ListUtils::listOf, ListUtils::merge));
    }
}

File: app/src/main/java/com/originb/inkwisenote2/common/ListUtils.kt
package com.originb.inkwisenote2.common

import java.util.function.Function

object ListUtils {
    @JvmStatic
    fun <T> listOf(vararg items: T?): List<T?> {
        return items.toList()
    }

    @JvmStatic
    fun <T> merge(a: List<T?>?, b: List<T?>?): List<T?> {
        val result = mutableListOf<T?>()
        a?.let { result.addAll(it) }
        b?.let { result.addAll(it) }
        return result
    }

    @JvmStatic
    fun <T, K> groupBy(list: List<T>, keyMapper: Function<T, K>): Map<K, List<T>> {
        return list.groupBy(keyMapper::apply).mapValues { it.value }
    }
}
